src/adapter/src/subscribe.rs

```diff
@@ -94,8 +107,105 @@ impl ActiveSubscribe {
+                                            start.0 == row.0
+                                                && compare_columns(
+                                                    order_by,
+                                                    &left_datums,
+                                                    &right_datums,
+                                                    || Ordering::Equal,
+                                                ) == Ordering::Equal
+                                        }));
+                                    let mut saw_new_row = false;
+                                    for (t, r, d) in group {
+                                        // We don't have to consider (k, v, -1), (k, v, +1) because
+                                        // the rows are already consolidated. For a key, if all we
+                                        // see is retractions we generate (k, null), otherwise we
+                                        // produce all values that k can take. If there are multiple
+                                        // live values for a given key it's impossible for a client
+                                        // to figure out when they are all deleted with ENVELOPE UPSERT.
+                                        if *d < 0 {
+                                            continue;
+                                        }
+                                        let mut packer = row_buf.packer();
+                                        let datums = datum_vec.borrow_with(r);
+                                        for idx in key_indices {
+                                            packer.push(datums[*idx]);
+                                        }
+                                        for idx in 0..self.arity {
+                                            if !key_indices.contains(&idx) {
+                                                packer.push(datums[idx]);
+                                            }
+                                        }
+
+                                        saw_new_row = true;
+                                        new_rows.push((*t, row_buf.clone(), 1));
+                                    }
+
+                                    if !saw_new_row {
+                                        // emit deletion
+                                        let mut packer = row_buf.packer();
+                                        let datums = datum_vec.borrow_with(&start.1);
+                                        for idx in key_indices {
+                                            packer.push(datums[*idx]);
+                                        }
+                                        for _ in 0..self.arity - key_indices.len() {
+                                            packer.push(Datum::Null);
+                                        }
+                                        new_rows.push((start.0, row_buf.clone(), -1));
```

Comment:
The diff here is hard coded to -1, and above to 1. Why do we not use `*d`? What happens if d isn't 1 or -1?

I commented in the design doc about possibly having a third value for mz_state of `duplicate` or `unknown` or something to indicate we detected a non-unique key, which we could use to signify to the user that the output is no longer dependable. WDYT?

Reply:
Changed to `*d`. tbh i didn't really consider this case closely, but yeah, when `d` isn't -1 or 1 then we have another key violation. I still haven't added the handling for when we have a key violation but it'll be needed here too

Reply:
added mz_state="key violation"